chroma-common/blockdevices/blockdevice_zfs.py

```diff
@@ -129,8 +124,19 @@ def lock_pool(self):
         while not self.lock.i_am_locking():
             try:
                 self.lock.acquire(self.LOCK_ACQUIRE_TIMEOUT)
+                with open(self.lock.lock_file, 'w') as f:
+                    f.writelines(str(self.lock.pid))
             except LockTimeout:
-                pass
+                # prune locks owned by non-existent processes
+                with open(self.lock.lock_file, 'r') as f:
+                    contents = f.readlines()
+
+                assert len(contents) == 1 and contents[0].isdigit(), \
+                    'unexpected contents of lockfile %s: %s' % (self.lock.lock_file, contents)
+
+                # validate pid holding lock exists
+                if not pid_exists(int(contents[0])):
```

Comment:
@tanabarr: Proper cleanup of resources doesn't have anything to do with utopias.

It is simply good practice that if you create a resource, such as a lockfile, that you make a reasonable effort to clean it up when you are done with it.  Catching signals and doing the cleanup is part of that reasonable effort.

It is true, that as a belt-and-suspenders strategy, that when finding a lockfile that you think might be an orphan and that has a PID it you should verify if the PID still exists before trying to decide if the lockfile is stale.  But that is **guidance** as to whether you should forcibly break the lock or not and should be used as a last-resort effort because PIDs are re-used and as such it's entirely possible that the PID that is in the lockfile is not the process that created it.

Moreover, the certainty of the staleness of such a lockfile goes down (dramatically IMHO) when you know reasonable lockfile cleanup is not being undertaken.  Put another way, when you do find a lockfile and you do think it's stale, the confidence that it is stale goes up considerably because you know they are being cleaned up on a best-effort basis and not being left around for newcomers to try to determine the status of them.  In effect, the occurrence of having to forcibly break a lock becomes very low and might even signal a bug.